### PR TITLE
Cloud overrides core

### PR DESCRIPTION
**Pull request by renanrodrigo**
Support generic overrides from the contract server.
Currently, the supported overrides are `series` and `cloud`, but more can be added in the future.

Further testing, mainly integration testing, can be performed once contracts starts applying those specific overrides with anything different than the regular `series` one that we had before. Right now, no flow has changed.

## Desired commit type
 - [ ] Squash and merge: Using the "Proposed Commit Message"
 - [ ] Create a merge commit and use "Proposed Commit Message"
 - [x] Rebase and merge, no merge commit, rely only on existing commit messages

## Checklist
 - [x] I have updated or added any unit tests accordingly
 - [x] I have updated or added any integration tests accordingly
 - [ ] I have updated or added any documentation accordingly

### CODE REVIEW

**Comment by renanrodrigo** [comment]
Yellow flag: while changing this function, I noticed that we'll call it around _4K TIMES_ during attach. For now, we need to be careful with side effects (I needed to cache the `cloud-id` call here) but refactoring this in the future seems sensible.

File: uaclient/util.py [hunk 1]
```diff
@@ -100,7 +102,43 @@ def object_hook(o):
...
+
+
+def _select_overrides(
+    entitlement: Dict[str, Any], series_name: str, cloud_type: str
+) -> Dict[int, Dict[str, Any]]:
+    overrides = {}
+
+    selector_values = {"series": series_name, "cloud": cloud_type}
+
+    series_overrides = entitlement.pop("series", {}).pop(series_name, {})
```
Comment (lucasmoura) [inline uaclient/util.py:124]: Aren't you modifying the global `entitlement` dict here by performing `pop`. I am just worried if might need that information later in the process

File: uaclient/tests/test_util.py [hunk 2]
```diff
@@ -474,57 +474,216 @@ def test_get_platform_info_with_version(
...
+                "obligations": {
+                    "new_obligation": True,
+                    "some_obligation": True,
+                },
+            }
+        }
 
-        assert expected == orig_access
+        util.apply_contract_overrides(orig_access)
+        assert orig_access == expected
```
Comment (lucasmoura) [inline uaclient/tests/test_util.py:686]: Should we add a test where we replace something with the selector ?

Comment (renanrodrigo) [reply uaclient/tests/test_util.py:686]: This test covers the case where we add new things and replace some others (like the `some_affordance`). Sorry, maybe I missed your point here

Comment (lucasmoura) [reply uaclient/tests/test_util.py:686]: My mistake, I missed that the list will be completely replaced. Just ignore this

File: uaclient/util.py (hunk 1 above)
Comment (orndorffgrant) [reply uaclient/util.py:124]: Right this aligns with the old implementation.

Not for this PR, but I think a worthwhile refactor would be to make this more "functional". So it wouldn't modify the argument and would return a new object with overrides applied

File: uaclient/util.py [hunk 3]
```diff
@@ -117,23 +155,30 @@ def apply_series_overrides(
...
-            # If the key already exists and is a dict, update that dict using
-            # the override
-            current.update(value)
-        else:
-            # Otherwise, replace it wholesale
-            orig_access["entitlement"][key] = value
+
+    overrides = _select_overrides(orig_entitlement, series_name, cloud_type)
+
+    for _weight, overrides_to_apply in sorted(overrides.items()):
```
Comment (orndorffgrant) [inline uaclient/util.py:172]: TIL `sorted` sorts dict_items by key with no extra params

File: uaclient/util.py [hunk 4]
```diff
@@ -100,7 +102,43 @@ def object_hook(o):
         return o
 
 
-def apply_series_overrides(
+def _get_override_weight(
```
Comment (orndorffgrant) [inline uaclient/util.py:105]: Unit tests just for this function might be useful and pretty easy to write

**Review by orndorffgrant — COMMENTED** [review]
Overall looks great!

I think we're missing a test for this scenario: There is a entitlement with two overrides. Override has just a series selector and override 2 has just a cloud selector. Both match - which one wins? (should be the cloud one from reading the code, but we should have a test)

File: uaclient/util.py (hunk 4 above)
Comment (renanrodrigo) [reply uaclient/util.py:105]: sure will add

**Comment by renanrodrigo** [comment]
@orndorffgrant The test Lucas commented (line 633) is covering the case when those overrides are different: we have series only, cloud only, and series+cloud - in this case, all are applied.
The previous test assures that when the override is the same, the highest weight wins.
Maybe I am missing the point on the scenario you mentioned - could you elaborate?

**Comment by renanrodrigo** [comment]
@lucasmoura I addresses Grant's comments and included an integration test assuring cloud overrides are applied

**Review by lucasmoura — APPROVED** [review]

**Pull request merged**